server/src/weaverbird/pipeline/pipeline.py

```diff
@@ -256,6 +256,74 @@ def remove_void_conditions_from_filter_steps(
     return final_steps
 
 
+def _remove_empty_elements(
+    data: dict[str, Any] | list[Any] | None
+) -> dict[str, Any] | list[Any] | None:
+    """
+    This should delete all empty arrays and empty dict
+    """
+    data_transformed: Any  # mypy
+    if isinstance(data, dict):
+        if (
+            len(
+                data_transformed := {
+                    k: _remove_empty_elements(v)
+                    for k, v in data.items()
+                    if v is not None and not len(v) == 0
+                }
+            )
+            == 0
+        ):
+            return None
+        return data_transformed
```

Comment:
I think this code can be improved:
* The outer `if` can be replaced by a truthiness check
* The return value of `_remove_empty_elements(v)` should be checked

```suggestion
        data_transformed = {
              k: cleaned
              for k, v in data.items()
              if (cleaned := _remove_empty_elements(v)) is not None
        }
        return data_transformed or None
```

A similar form can be used for the `list` case below